demos/multi_material/compositional_buoyancy/van_keken_1997.py

```diff
@@ -0,0 +1,276 @@
+epsilon = Constant(min_mesh_edge_length / 4)
+
+# Initialise level set as a smooth step function
+psi.interpolate((1 + tanh(signed_dist_to_interface / 2 / epsilon)) / 2)
+# -
+
+# We next define materials present in the simulation. Here, the problem is
+# non-dimensionalised and can be described by the product of the expressions for the
+# Rayleigh and buoyancy numbers, RaB, which is also referred to as compositional
+# Rayleigh number. Therefore, we provide a value for thermal and compositional Rayleigh
+# numbers to define our approximation.
+
+# +
+buoyant_material = Material(RaB=-1)  # Vertical direction is flipped in the benchmark
+dense_material = Material(RaB=0)
+materials = [buoyant_material, dense_material]
+
+Ra = 0  # Thermal Rayleigh number
+
+RaB = field_interface(
+    [psi], [material.RaB for material in materials], method="arithmetic"
+)  # Compositional Rayleigh number, defined based on each material value and location
+
+approximation = BoussinesqApproximation(Ra, RaB=RaB)
+# -
+
+# As with the previous examples, we set up a *Timestep Adaptor* for controlling the
+# time-step length (via a CFL criterion) as the simulation advances in time. We specify
+# the initial time, initial time step $\Delta t$, and output frequency (in time units).
+
+time_now = 0  # Initial time
+delta_t = Function(R).assign(1)  # Initial time step
+output_frequency = 10  # Frequency (based on simulation time) at which to output
+t_adapt = TimestepAdaptor(
+    delta_t, u, V, target_cfl=0.6, maximum_timestep=output_frequency
+)
+
+# This problem has a constant pressure nullspace, handled identically to our previous
+# tutorials.
+
+Z_nullspace = create_stokes_nullspace(Z)
+
+# Boundary conditions are specified next: no slip at the top and bottom and free slip
+# on the left and ride sides.
+
+stokes_bcs = {
+    bottom_id: {"u": 0},
+    top_id: {"u": 0},
+    left_id: {"ux": 0},
+    right_id: {"ux": 0},
+}
+
+# We now set up our output. To do so, we create the output file as a ParaView Data file
+# that uses the XML-based VTK file format. We also open a file for logging and
+# instantiate G-ADOPT geodynamical diagnostic utility.
+
+# +
+output_file = VTKFile("output.pvd")
+
+plog = ParameterLog("params.log", mesh)
+plog.log_str("step time dt u_rms entrainment")
+
+gd = GeodynamicalDiagnostics(z, T, bottom_id, top_id)
+material_area = material_interface_y * lx  # Quantity of tracked material
+entrainment_height = 0.2  # Height above which entrainment diagnostic is calculated
+# -
+
+# Here, we set up the variational problem for the Stokes and level-set systems. The
+# former depends on the approximation defined above, and the level-set system includes
+# both advection and reinitialisation.
+
+# +
+stokes_solver = StokesSolver(
+    z,
+    T,
+    approximation,
+    bcs=stokes_bcs,
+    nullspace=Z_nullspace,
+    transpose_nullspace=Z_nullspace,
+)
+
+subcycles = 1  # Number of advection solves to perform within one time step
```

Comment:
Provide some context for this? You use 1, but is this the recommendation for users too? What are the other options available and what are the trade-offs?

Reply:
This is used when the CFL for the energy system differs from that of the level-set advection. The latter should not exceed 0.6, but @stephankramer mentioned that the one for the energy system is usually higher. Subcycling allows multiple level-set advection steps without updating Stokes and energy in the meantime.